src/controllers/controllerengine.cpp

```diff
@@ -98,30 +98,48 @@ void ControllerEngine::callFunctionOnObjects(QList<QString> scriptFunctionPrefix
     }
 }
 
-/* -------- ------------------------------------------------------
-Purpose: Resolves a function name to a QScriptValue including
-            OBJECT.Function calls
-Input:   -
-Output:  -
--------- ------------------------------------------------------ */
-QScriptValue ControllerEngine::resolveFunction(const QString& function) const {
-    QHash<QString, QScriptValue>::const_iterator i =
-            m_scriptValueCache.find(function);
-    if (i != m_scriptValueCache.end()) {
-        return i.value();
-    }
+/* ------------------------------------------------------------------
+Purpose: Turn a snippet of JS into a QScriptValue function.
+         Wrapping it in an anonymous function allows any JS that
+         evaluates to a function to be used in MIDI mapping XML files
+         and ensures the function is executed with the correct
+         'this' object.
+Input:   QString snippet of JS that evaluates to a function,
+         int number of arguments that the function takes
+Output:  QScriptValue of JS snippet wrapped in an anonymous function
+------------------------------------------------------------------- */
+QScriptValue ControllerEngine::wrapFunctionCode(const QString& codeSnippet,
+                                                int numberOfArgs) {
+    QScriptValue wrappedFunction;
 
-    QScriptValue object = m_pEngine->globalObject();
-    QStringList parts = function.split(".");
+    QHash<QString, QScriptValue>::const_iterator i =
+            m_scriptWrappedFunctionCache.find(codeSnippet);
 
-    for (int i = 0; i < parts.size(); i++) {
-        object = object.property(parts.at(i));
-        if (!object.isValid()) {
-            break;
+    if (i != m_scriptWrappedFunctionCache.end()) {
+        wrappedFunction = i.value();
+    } else {
+        QScriptValue function = m_pEngine->evaluate(codeSnippet);
+        if (!syntaxIsValid(codeSnippet)) {
+            wrappedFunction = m_pEngine->evaluate(
+                "(function () { print('Syntax error in: "
+                + codeSnippet + "'); })");
```

Comment:
Like @rryan said about Mixxx getting out of the business of resolving the `this` object by hackishly parsing strings, I think Mixxx should move away from the business of handling strings of JavaScript at all. I would prefer to move towards a fully JS mapping option in which scripts manage their own input handler callbacks dynamically (as a side benefit mapping development would be nicer because the XML file would never have to be reloaded).